### ISSUE TITLE
[Feature Request] Show warning when an other user is working on it

### ISSUE THREAD

**Issue by xiaohutai** [bolt/bolt#7255]
One of our clients asked whether it is possible to see whether a different user else is already _editing_ a record; editing would mean to have the edit record open in a browser. Maybe with a (timed) lock/unlock?

**Comment by GawainLynch** [comment]
Locking … almost as hard and danger-fraught as cache invalidation. :wink:

**Comment by jkazimir** [comment]
### Danger ahead!

https://github.com/jkazimir/bolt-extension-edit-locking

Not quite working from the marketplace yet, as I'm pretty sure I messed up by requiring ratchet. That said, it's otherwise working proof-of-concept code utilizing a web socket server as a source of truth for edit page connections, granting an edit lock to the first to open any content, and disabling the form/save UI for any others. There are some risks/caveats noted on the readme there, but it should hopefully be hackable to fit a handful of possible business needs for it.

So to answer your clients question: is it _possible_? Yes. But as Gawain noted, there's quite a lot to consider, from potential unsaved data loss going at least as far down as client network connectivity.

**Comment by GawainLynch** [comment]
The requirement here is going to be adding `"cboden/ratchet": "^0.4"` to your project *root* `composer.json`.

Reason being, it has more than a few dependencies that would make people's sites behave … um … :wink: 

On a PHP 7.1 install **only** requiing `cboden/ratchet` the install list it:
```
cboden/ratchet            v0.4.1
evenement/evenement       v3.0.1
guzzlehttp/psr7           1.4.2
psr/http-message          1.0.1
ratchet/rfc6455           v0.2.3
react/cache               v0.4.2
react/dns                 v0.4.11
react/event-loop          v0.4.3
react/promise             v2.5.1
react/promise-timer       v1.2.1
react/socket              v0.8.7
react/stream              v0.7.6
symfony/http-foundation   v4.0.2
symfony/polyfill-mbstring v1.6.0
symfony/routing           v4.0.2
```

You're immediately going to have fun with Symfony 3.4 or 4.0 components conflicting in the autoloader, plus if anyone also uses React in their root package … :cat:

**Comment by SahAssar** [comment]
If anyone wants to take a whack at this in the future I recommend using SSE instead of websockets. SSE is very simple to implement in php and still allows for async server -> client communication after a page is loaded. Client -> server communication (for acquiring a lock) can easily be done via normal ajax.

**Comment by bobdenotter** [comment]
I have it from a reliable source that there will be an extension to improve this. It will not _block_ concurrent editing, but it _will_ warn editors if another editor has a page opened for editing.

**Comment by xiaohutai** [comment]
It's going to be the who-is-editing extension: https://github.com/nestordedios/bolt-extension-who-is-editing for simple checking; still needs some small tweaks, but it's getting there. Closing this for now.
